File: scripts/auto_gimbal.py
#!/usr/bin/env python3

import rospy
from math import pi, cos, sin, atan2, sqrt
import serial # used to exchange data with arduino controller
from rdk_msgs.msg import navigation
from rdk_msgs.msg import target


class SubscribeAndPublish:

    def __init__(self):
        self.X = 0
        self.Y = 0
        self.Xpoi = 200
        self.Ypoi = 0
        self.Heading = 0
        self.heading_offset = 15
        self.ser = serial.Serial()
        try:
            self.ser = serial.Serial(
                port='/dev/gimbal',
                baudrate = 115200,
                parity = serial.PARITY_NONE,
                stopbits = serial.STOPBITS_ONE,
                bytesize = serial.EIGHTBITS,
                timeout = 0.1 )
        except:
            rospy.logerr("[auto_gimbal] COM port error! Closing...")
            rospy.signal_shutdown("[auto_gimbal] COM port error!")

        self.navPub = rospy.Subscriber('navigation_data', navigation, self.navigation_callback)
        self.tarPub = rospy.Subscriber('target_data', target, self.target_callback)

        rospy.loginfo("Automatic gimbal control is ready!")
    # End of __init__()

    def pi2pi(self, angle):
        while abs(angle) > 180:
            angle = angle - 360 * angle / abs(angle)
        return angle

    def bound(self, value, mi, ma):
        upper = max([mi, value])
        lower = min([upper, ma])
        return lower

    def navigation_callback(self, msg):
        self.X = msg.X
        self.Y = msg.Y
        self.Heading = msg.heading

    def target_callback(self, msg):
        self.Xpoi = msg.Xpoi
        self.Ypoi = msg.Ypoi

    def generate_controls(self):
        peleng = atan2(self.Ypoi - self.Y, self.Xpoi - self.X)
        gimbal_heading = round((peleng - self.Heading) * 180.0 / pi) - self.heading_offset
        gimbal_heading = self.pi2pi(gimbal_heading)
        distance = sqrt((self.Ypoi - self.Y)**2 + (self.Xpoi - self.X)**2)
        optimal_distance = 7  # meters
        gimbal_pitch = -round((distance - optimal_distance) * 3)
        rospy.loginfo("Debug: distance: {:.1f}".format(distance - optimal_distance))
        rospy.loginfo("Debug: g_pit: {:.1f}".format(gimbal_pitch))
        gimbal_heading = self.bound(gimbal_heading, -110, 110)
        gimbal_heading = int(gimbal_heading / 2)
        gimbal_pitch = self.bound(gimbal_pitch, -40, 20)
        self.ser.flush()
        self.ser.write("campos {} {} \n".format(gimbal_heading, gimbal_pitch).encode())
        rospy.loginfo("Sent [campos {} {}]".format(gimbal_heading, gimbal_pitch))
    # End of generate_controls()

# End of SAP class


def main():
    rospy.init_node('auto_gimbal_node')
    SAP = SubscribeAndPublish()

    rate = rospy.get_param("~rate", 0.2)
    r = rospy.Rate(rate)  # Hz
    while not rospy.is_shutdown():
        # Main cycle
        SAP.generate_controls()
        r.sleep()
        #try:
        #    code, msg, val = m.getSystemState(caller_id)
        #except:
        #    break
    rospy.loginfo("[auto_gimbal_node] Closing...")
    SAP.ser.close()
# End on main()


if __name__ == "__main__":
    main()

File: src/auto_gimbal.cpp
#include "ros/ros.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include <cstring>
#include <iostream>
#include <rdk_msgs/navigation.h>
#include <rdk_msgs/target.h>
#include "CppSerial.cpp"
#include <ros/master.h>
#include <sstream>


class SubscribeAndPublish
{
private:
  CppSerial ser;

  float X = 0;
  float Y = 0;
  float Xpoi = 200;
  float Ypoi = 0;
  float Heading = 0;

  ros::NodeHandle nh;
  ros::Subscriber navigationSub;
  ros::Subscriber targetSub;


public:
    SubscribeAndPublish() // This is the constructor
    {
      char portname[] = "/dev/ttyUSB0";
      ser.Open(portname, 115200);

      targetSub = nh.subscribe<rdk_msgs::target>("target_data", 5, &SubscribeAndPublish::target_callback, this);
      navigationSub = nh.subscribe("navigation_data", 5, &SubscribeAndPublish::navigation_callback, this);

    }

    int bound(int value, int mi, int ma)
    {
      int upper = std::max({mi, value});
      int lower = std::min({upper, ma});
      return lower;
    }

    int convertZeroPadedHexIntoByte(char *dataset, unsigned char *bytearray)
    {
    	int le = strlen(dataset);

      for(int j = 0; j < le; j++)
      {
        bytearray[j] = (dataset[j] < 0)?(dataset[j] + 256):dataset[j];
        //bytearray[j] = static_cast<unsigned char>(dataset[j]);
      }
      return 0;
    }

    void navigation_callback(const rdk_msgs::navigation msg)
    {
      X = msg.X;
      Y = msg.Y;
      Heading = msg.heading;
      return;
    }

    void target_callback(const rdk_msgs::target msg)
    {
      Xpoi = msg.Xpoi;
      Ypoi = msg.Ypoi;
      return;
    }

    void generateControl()
    {
      float peleng = atan2(Ypoi - Y, Xpoi - X);
      int gimbal_heading = round((peleng - Heading) * 180.0 / M_PI);
      float distance = sqrt(pow(Ypoi - Y, 2) + pow(Xpoi - X, 2));
      float optimal_distance = 5;  // meters
      int gimbal_pitch = -round((distance - optimal_distance) * 3);
      gimbal_heading = bound(gimbal_heading, -90, 90);
      gimbal_heading = (int) (gimbal_heading / 2);
      gimbal_pitch = bound(gimbal_pitch, -40, 20);

      char MSG[32] = {'\0'};
      memset(MSG, 0, sizeof(MSG));
      sprintf(MSG, "campos %d %d \n", gimbal_heading, gimbal_pitch);

      ser.FlushTransmit();
      //ser.Send((unsigned char*)MSG, sizeof(MSG));
      std::string das = std::string(MSG);
      ROS_INFO("Sent: %s\n", das.c_str());
      ser.SendString(das);
    }


};//End of class SubscribeAndPublish

int main(int argc, char** argv)
{
    ros::init(argc, argv, "auto_gimbal");

    // Handle ROS communication events
    SubscribeAndPublish SAPObject;

    // For cycling operation use
    ros::Rate rate(0.25); // ROS Rate at 20 Hz
    bool master_is_alive = true;
    while (ros::ok() && master_is_alive)
    {
      SAPObject.generateControl();

      if (!ros::master::check())
      {
        ROS_WARN_THROTTLE(5, "ROS Master is not found, shutting down...");
        master_is_alive = false;
      }
     ros::spinOnce();
     rate.sleep();
    }

  	return 0;
  }
